Concretizer: put build->run deps in parent unification set (fixes #46034)

muffgaga opened the pull request:
In #46034, a package X build-depends on a `build-tools`-tagged package, e.g., `py-setuptools`, as well as another package Y that itself depends (build+run) on the same `build-tools` tagged package. In this case, the concretizer puts the possibly two versions of the build tool into separate unification sets, but the run-type dependency requires the
build environment of X to have both versions in its environment — which is illegal.
With this fix, the concretizer puts the two build tool nodes into a new common unification set if there is a build→run-type dependency chain on the build tool ⇒ spack now reports the expected error about not being able to find a common concretization of the build tool.

muffgaga commented:
BTW. the fail in `lib/spack/spack/test/concretize.py::TestConcretizeSeparately::test_two_setuptools` looks valid to me… there are two `py-setuptools` versions that will end up in `py-shapely`'s build environment (I'm trying to get `spack build-env` working for the `duplicates.test` repo but have failed so far :/).

muffgaga commented:
@elenimath suggests this restructuring of the unification sets:
```
diff --git a/lib/spack/spack/solver/concretize.lp b/lib/spack/spack/solver/concretize.lp
index 18c82474c9..923c195f35 100644
--- a/lib/spack/spack/solver/concretize.lp
+++ b/lib/spack/spack/solver/concretize.lp
@@ -95,12 +95,19 @@ unify(SetID, PackageName) :- unification_set(SetID, node(_, PackageName)).
 unification_set("root", PackageNode) :- attr("root", PackageNode).
 unification_set(SetID, ChildNode) :- attr("depends_on", ParentNode, ChildNode, Type), Type != "build", unification_set(SetID, ParentNode).
 
-unification_set(("build", node(X, Child)), node(X, Child))
+unification_set(("build", ParentNode), node(X, Child))
   :- attr("depends_on", ParentNode, node(X, Child), Type),
      Type == "build",
      multiple_unification_sets(Child),
      unification_set(SetID, ParentNode).
 
+unification_set(("build", ParentNode), node(X, Child))
+  :- attr("depends_on", ParentNode, NonDuplicateChild, "build"),
+     not multiple_unification_sets(NonDuplicateChild),
+     attr("depends_on", NonDuplicateChild, node(X, Child), "run"),
+     multiple_unification_sets(Child),
+     unification_set(_, ParentNode).
+
 unification_set("generic_build", node(X, Child))
   :- attr("depends_on", ParentNode, node(X, Child), Type),
      Type == "build",
```
⇒ that looks like the complete solution to me.

muffgaga committed:
fix: TestConcretizeSeparately tests involving multiple unification sets

This introduces `-fixed` packages for py-shapely and py-floating which
depend on a `-fixed` py-numpy that does NOT run-depend on py-setuptools.
If we allow run deps of build deps, it causes the tool to appear multiple
times in the build environment ⇒ fail.

Change-Id: I1ef1fa0b8636c23626c7e2ea521d3b5438f97d0a

haampie commented:
Brave to take on this concretizer stuff ;) @alalazo will take a look

alalazo commented:
@muffgaga I was looking at the issue, and I think this is what I would apply to `develop` as a change:
```diff
diff --git a/lib/spack/spack/solver/concretize.lp b/lib/spack/spack/solver/concretize.lp
index 18c82474c9..51dbba68c8 100644
--- a/lib/spack/spack/solver/concretize.lp
+++ b/lib/spack/spack/solver/concretize.lp
@@ -95,7 +95,7 @@ unify(SetID, PackageName) :- unification_set(SetID, node(_, PackageName)).
 unification_set("root", PackageNode) :- attr("root", PackageNode).
 unification_set(SetID, ChildNode) :- attr("depends_on", ParentNode, ChildNode, Type), Type != "build", unification_set(SetID, ParentNode).
 
-unification_set(("build", node(X, Child)), node(X, Child))
+unification_set(("build", ParentNode), node(X, Child))
   :- attr("depends_on", ParentNode, node(X, Child), Type),
      Type == "build",
      multiple_unification_sets(Child),

```
There are two tests failing locally for me, I'll try to check them. 

As you said in https://github.com/spack/spack/pull/46106#issuecomment-2317637177 I think we should have two tests.

1. If it's the root to have a pure build dep on `py-setuptools` and on something that has a run dependency on `py-setuptools`, the `py-setuptools` node should be unique and both edges should point to it (if the problem is solvable)
2. In the opposite case, where the root has the build + run dep on `py-setuptools`  and on something that has only a build dep to py-setuptools, then we should allow for two `py-setuptools` nodes, when required.

alalazo commented:
@muffgaga Go with this change you suggested, thanks https://github.com/spack/spack/pull/46106#issuecomment-2317678517 !

alalazo commented:
This diff instead:
```diff
diff --git a/var/spack/repos/duplicates.test/packages/py-numpy/package.py b/var/spack/repos/duplicates.test/packages/py-numpy/package.py
index 42cb2aecc7..e0ca7dfa32 100644
--- a/var/spack/repos/duplicates.test/packages/py-numpy/package.py
+++ b/var/spack/repos/duplicates.test/packages/py-numpy/package.py
@@ -16,5 +16,5 @@ class PyNumpy(Package):
     version("1.25.0", md5="0123456789abcdef0123456789abcdef")
 
     extends("python")
-    depends_on("py-setuptools@=59", type=("build", "run"))
+    depends_on("py-setuptools@=59", type="build")
     depends_on("gmake@4.1", type="build")
diff --git a/var/spack/repos/duplicates.test/packages/py-shapely/package.py b/var/spack/repos/duplicates.test/packages/py-shapely/package.py
index 091f8c9938..f5809d3341 100644
--- a/var/spack/repos/duplicates.test/packages/py-shapely/package.py
+++ b/var/spack/repos/duplicates.test/packages/py-shapely/package.py
@@ -15,4 +15,4 @@ class PyShapely(Package):
 
     extends("python")
     depends_on("py-numpy", type=("build", "link", "run"))
-    depends_on("py-setuptools@=60", type="build")
+    depends_on("py-setuptools@=60", type=("build", "run"))

```
should make the current tests pass. Then we'd just need a test for the case that should fail.

muffgaga commented:
@alalazo Thanks for looking into this. I'm currently creating another test that checks some "build-tool with its own build deps" case, and I'm not yet sure that everything works there (we don't catch it in the tests, because `py-setuptools` doesn't have deps in those). I also added "fixed" tests (and the inverse one) → it's probably better style to keep the package names the same and add variants for `+rundep` or something like that, right?
(And I'll update the PR according to your suggestions.)

elenimath commented:
@alalazo about https://github.com/spack/spack/pull/46106#issuecomment-2317678517, I just assumed that smaller unification sets were better performance-wise, but I'm not sure what the actual impact would be :)

muffgaga commented:
> @alalazo about [#46106 (comment)](https://github.com/spack/spack/pull/46106#issuecomment-2317678517), I just assumed that smaller unification sets were better performance-wise, but I'm not sure what the actual impact would be :)

I did a quick `spack solve --timers -UtI nest` on my laptop (`develop` from this morning:

@alalazo's nice-and-clean patch:
```
    setup          3.699s
    load           0.460s
    ground         5.080s
    solve         10.756s
    total         20.489s
```

@elenimath yours :P
```
    setup          3.757s
    load           0.482s
    ground         4.320s
    solve          6.183s
    total         15.235s
```

muffgaga commented:
Changes
* adapted change for `concretize.lp` (and fix author info)
* duplicates.test repo: PyNumpy has now a `rundeps` variant but defaults to build-only dependency on `py-setuptools`
* duplicates.test repo: PyShapely now run-depends on `py-setuptools` (to check that this is possible)

alalazo commented:
Looks good, thanks. I'll do some performance benchmark asap, and post it.

muffgaga commented:
I added another test that shows that there's a problem with "strategy: full".

The test should cover this case here:
```
#  all build deps
o compiled-tool@1.0
|\
o | seemake@4.0
o | gmake@4.1
 /
o compiled-intermediate-tool@3.0
o seemake@3.0
o gmake@3.0
```

But in `strategy: full` when I look at the `self.{_link_run,_total_build,_direct_build}` members, `gmake` is there only once, i.e. `max_dupes` for it is set to one → fails to concretize.
In `strategy: minimal` it does not happen because `gmake` is a build tool and a max_dupes of 2 is set for it.

elenimath committed:
test: add build->run dep regression test in TestConcretizeSeparately

This introduces a regression test to ensure that the concretizer fails
when trying to solve a spec where we have a build tool in two versions and
the "lower" (in the DAG) version being also a run-dependency. If this
would be allowed, we could end up having multiple paths to different tool
versions/variants in the build env.

Change-Id: I050efc9471d89580f5546a1c56ab419891bcc792

muffgaga commented:
I removed the "full"-strategy related things from this PR as it is an unrelated issue → @elenimath did a deeper investigation and will provide a fresh PR (including the initially failing PR and maybe also a fix :) ).

alalazo commented:
I just benchmarked on:

* **Spack:** 0.23.0.dev0 (1f935ac356b8b7427d72b9226f12bf75e74df120)
* **Python:** 3.11.5
* **Platform:** linux-ubuntu20.04-icelake

[radiuss-pr.csv](https://github.com/user-attachments/files/16844416/radiuss-pr.csv)
[radiuss-develop.csv](https://github.com/user-attachments/files/16844417/radiuss-develop.csv)
[radiuss.txt](https://github.com/user-attachments/files/16844418/radiuss.txt)

![radiuss](https://github.com/user-attachments/assets/e868709a-ba5a-4031-a1a1-5ed73b8240c5)

Using unification sets as "build environments" per node comes with a significant performance hit. I'll check later the patch in https://github.com/spack/spack/pull/46106#issuecomment-2317678517, but it seems to me that this one is overly specific to this case (e.g. why only run dependencies and not link dependencies?). 

I'll try to think how to optimize this use case :thinking:

elenimath commented:
> e.g. why only run dependencies and not link dependencies?

true, it should be `Type != "build"` instead

muffgaga commented:
> > e.g. why only run dependencies and not link dependencies?
> 
> true, it should be `Type != "build"` instead

(`"run"` vs. `Type != build` doesn't seem to affect solving time much.)

> Using unification sets as "build environments" per node comes with a significant performance hit. I'll check later the patch in https://github.com/spack/spack/pull/46106#issuecomment-2317678517, […]

Some single runs of `glvis` (which shows the hardest impact in your plots) and see for the total times (most of the difference is in the solving time, as your plots also show); this is all ± 1s depending on what my browser does in parallel ;)

* develop: 22.5s
* @alalazo's suggested fix: 38.5s (i.e. +70%)
* @elenimath's suggested fix (modified, w/ `Type != "build"`): 28s (i.e. +25%)

haampie commented:
I think it was sensible to leave out type=link. Link deps roughly correspond to local search paths or no search at all, run deps have their global `*PATH`s, which is the issue being solved.

alalazo commented on var/spack/repos/duplicates.test/packages/py-shapely/package.py:
```diff
@@ -15,4 +15,4 @@ class PyShapely(Package):
     extends("python")
     depends_on("py-numpy", type=("build", "link", "run"))
-    depends_on("py-setuptools@=60", type="build")
+    depends_on("py-setuptools@=60", type=("build", "run"))
```
I think we should avoid this change:
```suggestion
    depends_on("py-setuptools@=60", type="build")
```
Without it, we model the use case in the issue correctly. Besides, with the change, all unit tests pass also on develop (without, instead, 2 tests would fail on develop).

alalazo reviewed (changes requested):
I'm currently looking at performance vs. correctness. 

I think the most correct change is the one in https://github.com/spack/spack/pull/46106#issuecomment-2318106245 but it's really slow (due to an explosion of the rules for unification sets).

I'm fine with the heuristic in https://github.com/spack/spack/pull/46106#issuecomment-2317678517. The limitation there is that it doesn't account for transitive dependencies of `NonDupeNode` in the link/run subdag. It's also slower than `develop` due to the same reason as above, but at a lower scale.

TL;DR I'm looking for some alternative heuristic to avoid slowing down the solve. If we can't find it, I'm fine with https://github.com/spack/spack/pull/46106#issuecomment-2317678517. The request for change is for the comment on the test. Can you double check it?

elenimath committed:
Concretizer: build-tools with build->run dep chain in unification set (fixes #46034)

* Each parent gets its own "build environment" where the build deps are
  unified.
* Things that don't allow multiple nodes are also thrown into the
  "generic_build" unification

In #46034, a package X build-depends on a `build-tools`-tagged package,
e.g., `py-setuptools`, as well as another package Y that itself depends
(build+run) on the same `build-tools` tagged package. In this case,
the concretizer puts the possibly two versions of the build tool into
separate unification sets, but the run-type dependency requires the
build environment of X to have both versions in its environment — which
is illegal.
With this fix, the concretizer puts the two build tool nodes into a new
common unification set if there is a build→run-type dependency chain on
the build tool ⇒ spack now reports the expected error about not being
able to find a common concretization of the build tool.

Change-Id: I99257b225602c21d4f270fe16e68c1f8bd4b1eba
